core/src/main/scala/reconciler/Reconciler.scala

```diff
@@ -119,6 +119,118 @@ abstract class Reconciler[S <: CustomResource[_, _], T <: ObjectResource] {
+
+  case class Resource[O <: ObjectResource](watch: akka.stream.scaladsl.Source[WatchEvent[O], _], list: List[O])
+  case class ResourcesState(source: akka.stream.scaladsl.Source[WatchEvent[S], _],
+                            target: akka.stream.scaladsl.Source[WatchEvent[T], _],
+                            state: Map[CacheEntryKey, CacheEntry])
+
+  def buildResource[O <: ObjectResource](implicit context: RequestContext,
+                                         format: Format[O],
+                                         listFormat: Format[ListResource[O]],
+                                         resourceDefinition: ResourceDefinition[O],
+                                         ec: ExecutionContext): Future[Resource[O]] = {
+    context.list[ListResource[O]]().map { l =>
+      Resource[O](
+        context.watchAllContinuously[O](Some(l.resourceVersion)),
+        l.items
+      )
+    }
+  }
+
+  def buildState(sources: List[S], target: List[T]): Map[CacheEntryKey, CacheEntry] = {
+    val initKeys = sources.foldLeft(Map.empty[CacheEntryKey, CacheEntry]) {
+      case (acc, s) =>
+        val key = CacheEntryKey(name = s.name, namespace = s.namespace, kind = s.kind)
+        acc + (key -> CacheEntry(s, List.empty))
+    }
+
+    target.foldLeft(initKeys) {
+      case (acc, t) =>
+        val owner = t.metadata.ownerReferences.head
+        val sourceKey = CacheEntryKey(name = owner.name, namespace = t.namespace, kind = owner.kind)
+
+        acc.get(sourceKey).fold(acc)(s => acc + (sourceKey -> s.copy(ts = t :: s.ts)))
+    }
+  }
+
+  def buildResourcesState(implicit context: RequestContext,
+                          sourceFormat: Format[S],
+                          sourceListFormat: Format[ListResource[S]],
+                          sourceResourceDefinition: ResourceDefinition[S],
+                          targetFormat: Format[T],
+                          targetListFormat: Format[ListResource[T]],
+                          targetResourceDefinition: ResourceDefinition[T],
+                          ec: ExecutionContext): Future[ResourcesState] = {
+
+    for {
+      source <- buildResource[S]
+      target <- buildResource[T]
+    } yield
+      ResourcesState(
+        source.watch,
+        target.watch,
+        buildState(source.list, target.list)
+      )
+
+  }
+
+  def scanner(acc: Cache, x: WatchEvent[_ >: T with S <: ObjectResource]): Cache = {
+    val `WatchEvent[S]` = TypeCase[WatchEvent[S]]
+    val `WatchEvent[T]` = TypeCase[WatchEvent[T]]
+
+    x match {
+      case `WatchEvent[S]`(we) =>
+        val key = CacheEntryKey(name = we._object.name, namespace = we._object.namespace, kind = we._object.kind)
+        val targets = acc.cache.get(key).fold(List.empty[T])(_.ts)
+        val cache = acc.cache + (key -> CacheEntry(we._object, targets))
+
+        we._type match {
+          case EventType.ADDED | EventType.MODIFIED => Cache(List(Modified(we._object, targets)), cache)
```

Comment:
probably be good to have a separate add and modified for this.

Reply:
Just separate cases for `ADDED` and `MODIFIED` or introducing a new `Added` to the `Event` ADT?

Reply:
We thought about separate `Added` and  `Modified` `Event`s but couldn't think of a use call in the reconciler. Because the event had the source and Target list added and modified result in the same actions.

I'd be really interested in your thoughts here. Is out logic correct?

Reply:
We thought about separate `Added` and  `Modified` `Event`s but couldn't think of a use call in the reconciler. Because the event had the source and Target list added and modified result in the same actions.

I'd be really interested in your thoughts here. Is out logic correct?

Reply:
Similarly, have not found a use case for it in the implementations I have done. Though I probably more worried the uses that i've not come across... but in either case lets go with this.